### DISCUSSION TITLE
I'm interested in beta testing the Plasma/THC Plugin

### DISCUSSION THREAD

**Discussion by bLouChip** [grblHAL/plugins#18] [Q&A] [answered]
Greetings,
I'm interested in beta testing the Plasma/THC Plugin.  However I'm not a current user of grblHAL; I would need to purchase the controller board and any peripherals such as a display panel is so needed.  My CNC machine is MillRight Mega V XL.  
https://millrightcnc.com/product/millright-cnc-mega-v-router-bundle/
It's running the stock Arduino mega2560 board that MR ships with it, as well as the stock NEMA 23? motor drivers.  MR forked grbl in 2018-'19 to make a custom grbl v1.1i with 4th (A) axis support. My machine is XYZA axis and setup to run 3 processes: routing/milling, plasma, and laser. I use SheetCAM for my plasma toolpath gen, and I already have a custom .scpost post processor so it's not a problem for me to clone it and modify to output the necessary THC gcodes defined for your plasma plugin. I use LightBurn for laser CAD/CAM. I use Aspire CAD/CAM for routing/milling which also has post processor I can modify.  When not using LightBurn as the gcode sender (laser process only), then I use UGS V2.0.12. I'm not sure how this factors into the display info produced by Plasma/THC plugin unless there is a hw display connected to grblHAL, or you'd have to recommend another gcode sender for me to use. 

In plasma process, I use a Proma THC SD which is designed to directly control Z with STEP and DIR inputs to the motor driver.  The unit I don't have, the Proma THC 150, is designed to work with your Plasma/THC Plugin in $350=2 mode.  If we can't adapt a new mode for the THC SD by monitoring its output for STEP and DIR then grblHAL behaving as if its a 150 model (this is my preference), then I'd have to use $350=1 mode, which concerns me due to bringing the arc voltage signal into the control box, its very noisy electrically.  However, I like what I see in your Plasma/THC plugin in terms of 1) grblHAL controlling the Z motor exclusively and thus being able to keep Z machine zero and 2) the Velocity Anti Dive logic.   Interestingly, I recently built a TTL circuit with a couple of 555 timers, AND gates, and a S/R flipflop to perform the same VAD function (I called it THC Z Step Suppressor circuit), using Spindle PWM signal as the velocity indicator/signal and M7 as a circuit enable/disable signal.  It's all explained here: https://millrightcnc.proboards.com/thread/3563/plasma-router-custom-build-frame . As you will read in the referenced blog, I went through the gauntlet of EMI/RFI issues with CNC plasma process to get it working very well and reliable now, so if I have to do that again with grblHAL hardware, I can and I will, not an issue. 

Although I tinker with logic and analog circuits, micro controllers, software, and stuff mechanical, I'm not very interested this year in building the grblHAL board; I'd consider it, but only after considering the purchase of a fully built board that I could just hook up to my motor drivers, limit sws, pause, resume, safetydoor, and THC.  My testing would involve not just plasma process, but laser and milling as well, but plasma and laser being the most used. 

Can you point me to a grblHAL board fully assembled and compatible (HW) with Plasma/THC and LightBurn Laser plugins, for purchase ? 

Cheers,
Lou

**Comment by bLouChip** [comment]
ok, so upon further review... I missed this version the first time I searched, there is another version of this board that was etched in error, noted as such, and selling for less. 
https://www.tindie.com/products/philba/grblhal-bob-unkit-for-teensy-41-t41u5xbb/
btw - I also just noticed that the mfg, "Phil", will pre-flash the board with grblHAL, hopefully with the plugins I mention complied in. 

Anyway, could one of the developers please confirm this model works with Plasma/THC and LB/Laser plugins.  I'll check with "Phil" also before ordering.  Thanks much.

**Comment by terjeio** [comment, marked as answer]
A lot to digest here, I will need some time to go through your impressive work. With your help I believe it should be possible to make the plugin work well.

> I'm not sure how this factors into the display info produced by Plasma/THC plugin unless there is a hw display connected to grblHAL, or you'd have to recommend another gcode sender for me to use.

Currently there is no (direct) display support at all in grblHAL, but it is possible to provide data - either via a plugin or via the MPG mode that many drivers support. It is recommended that a display is controlled by its own MCU that uses data from the controller for rendering. [This project ](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack) uses MPG mode to extract data (mainly from the real time report) and [this has a custom interface](https://github.com/Expatria-Technologies/RT_Jog_Controller) that uses I2C for communication. 
ioSender has support (untested) for parsing and displaying data from the THC specific real time report extensions so that could be a candidate as well.

> I also just noticed that the mfg, "Phil", will pre-flash the board with grblHAL

I don't think so, you will have to source the Teensy 4.1 yourself and flash it. He provides precompiled firmware, but since the [Web Builder](http://svn.io-engineering.com:8080/) was launched it is easier to build your own firmware to spec. The plasma plugin is not yet available from the Web Builder, it is no big deal for me to add it.

I cannot confirm that the Plasma/THC plugin will work as I have only done simple bench testing - more work is certainly needed before it is ready for prime time. Beta testing implies that there will likely be changes...
The LaserBurn plugin is pretty generic and can be used with most drivers.

[This list](https://github.com/grblHAL/Controllers) contains links to some of the other controllers supported by grblHAL, but be aware that (IIRC) only the iMXRT1062 has driver support for outputting steps for the Z motor. Other drivers can be updated depending on whether they have the needed resources (at least a free timer).

**Comment by bLouChip** [comment]
@terjeio , thank you for the quick answer.   From my brief read of the references, I believe [ioSender ](https://github.com/terjeio/ioSender) will do just fine.  The other hw/sw options are impressive though.  You mention in the ioSender readme that it supports up to 9 axes, so even though I don't see it on the DRO, I suspect the A axis will appear if grblHAL is so configured.  

I'll keep you posted if/when I progress to a grblHAL alpha/beta testing project, in the office first, in the shop on the CNC second. 
thanks again,
Lou
